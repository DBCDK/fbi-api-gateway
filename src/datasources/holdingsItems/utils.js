// Helper function to construct the URL dynamically
export function buildPath(
  url,
  { agencyId, bibliographicRecordId, trackingId, itemId }
) {
  const basePath = [url, agencyId, bibliographicRecordId, itemId]
    .filter(Boolean)
    .join("/");
  const query = trackingId
    ? `?trackingId=${encodeURIComponent(trackingId)}`
    : "";
  return `${basePath}${query}`;
}
