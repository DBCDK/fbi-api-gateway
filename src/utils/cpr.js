/**
 * Check for validity of a cpr
 *
 * @param cpr
 * @returns {boolean|Date|boolean}
 */
export function isValidCpr(cpr) {
  return isNumeric(cpr) && isValidDate(cpr) && cpr.length === 10;
}

/**
 *
 * @param n
 * @returns {boolean}
 */
function isNumeric(n) {
  return !isNaN(parseFloat(n)) && isFinite(n);
}

/**
 * Check if first 6 digits contains a valid date
 * @param ddmmyy
 * @returns {Date|boolean}
 */
function isValidDate(ddmmyy) {
  const mm = ddmmyy.substr(2, 2) || "00";
  const d = new Date(ddmmyy.substr(4, 2), mm - 1, ddmmyy.substr(0, 2));
  return d && d.getMonth() + 1 === parseInt(mm, 10);
}
