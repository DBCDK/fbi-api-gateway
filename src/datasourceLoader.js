import DataLoader from "dataloader";

import { log } from "dbc-node-logger";
import { withRedis, clearRedis } from "./datasources/redis.datasource";
import { createFetchWithConcurrencyLimit } from "./utils/fetchWithLimit";
import { getFilesRecursive } from "./utils/utils";
import config from "./config";

/*
  This is how we time parallel requests to a single datasource.
  
  Request A
  ------------
          Request B
          ------------------
                                      Request C
                                      -----------
  Total AB
  --------------------------
  
  The totaltime for the requests should NOT be:
  Request A + Request B + Request C
  
  But instead:
  Total AB + Request C
*/

export const trackMe = () => {
  return {
    start: function (uuid) {
      this.uuid = uuid;
      return this;
    },
    createTracker: function (name) {
      // If this datasource has not been tracked yet, initialize its tracking object.
      if (!this.trackObject[name]) {
        this.trackObject[name] = {
          overlappingRequestCount: 0, // Number of ongoing requests for this datasource.
          count: 0, // Total number of elements retrieved from datasource
          started: 0, // Timestamp when the first concurrent request starts.
          time: 0, // Total time taken for all requests so far.
          sequentialTime: 0, // If requests were to be run sequentially, this is the total time
          batches: 0, // Dataloader batches multiple requests, into a single request when possible
        };
      }
    },

    // Function to indicate the beginning of a datasource call.
    begin: function (name, count) {
      this.createTracker(name);

      const beginTime = performance.now();

      const trackObject = this.trackObject[name];

      // Increment the count of ongoing requests.
      trackObject.overlappingRequestCount++;

      // Increment total number of elements retrieved from datasource
      trackObject.count += count;
      trackObject.batches++;

      // Function to indicate the end of an datasource call.
      return function end() {
        const endTime = performance.now();
        const duration = Math.round(endTime - beginTime);
        // Decrement the count of ongoing requests.
        trackObject.overlappingRequestCount--;

        // If all ongoing requests are complete, calculate the time taken.
        if (trackObject.overlappingRequestCount === 0) {
          trackObject.time += duration;
        }

        trackObject.sequentialTime += duration;
      };
    },

    cacheMiss: function (name, count) {
      this.createTracker(name);

      this.trackObject[name].cacheMiss =
        (this.trackObject[name].cacheMiss || 0) + count;
    },
    uuid: null,
    trackObject: {},
    getMetrics: function () {
      const res = {};

      Object.entries(this.trackObject).forEach(([key, val]) => {
        res[key] = {
          count: val.count,
          time: val.time,
          cacheMiss: val.cacheMiss,
          avg_ms: Math.round(val.sequentialTime / val.count),
        };
      });

      return res;
    },
  };
};

// Find all datasources in src/datasources
export const datasources = getFilesRecursive(`${__dirname}/datasources`)
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status, testLoad } = require(file.path);
    if (!load && !batchLoader) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    return {
      testLoad,
      load,
      batchLoader,
      name,
      options,
      statusChecker: status ? () => status(load) : null,
    };
  })
  .filter((func) => !!func);

const nameToDatasource = {};
datasources.forEach((ds) => (nameToDatasource[ds.name] = ds));

log.debug(
  `found ${datasources.length} datasources, ${datasources
    .map((datasource) => datasource.name)
    .join(", ")}`
);

/**
 * This will initialize a single dataloader.
 * It makes sure a context (with a concurrency limited fetcher), is injected
 * into the batchloader function.
 *
 * It sets up Redis for the dataloader
 *
 * It collects performance metrics for the datasource.
 */
function setupDataloader(
  { name, load, options, batchLoader, testLoad },
  context
) {
  const testModeActivated = context?.testUser && testLoad;
  const loadImpl = testModeActivated ? testLoad : load;

  let batchLoaderWithContext = batchLoader
    ? (keys) => batchLoader(keys, context)
    : (keys) => Promise.all(keys.map((key) => loadImpl(key, context)));

  if (!testModeActivated && options?.redis?.prefix && options?.redis?.ttl) {
    batchLoaderWithContext = withRedis(batchLoaderWithContext, {
      ...options.redis,
      ...context,
      datasourceName: name,
    });
  }

  async function batchLoaderWithTiming(keys) {
    const end = context?.track?.begin?.(name, keys.length);

    try {
      return await batchLoaderWithContext(keys);
    } catch (err) {
      // Wraps the load call in a try catch, in order to capture the
      // stack trace before dataloader is called. Otherwise we would
      // be limited to see the stack trace from the data sources, but
      // not the GraphQL resolvers.
      // https://brunoscheufler.com/blog/2021-07-25-improving-dataloader-stack-traces
      if (err instanceof Error && err.stack) {
        const currentStack = { stack: "" };
        Error.captureStackTrace(currentStack, batchLoaderWithTiming);
        err.stack += currentStack.stack.replace("Error", "\nBefore DataLoader");
      }
      throw err;
    } finally {
      end?.();
    }
  }

  const loader = new DataLoader(batchLoaderWithTiming, {
    // If key is an object, we stringify
    // to make it useful as a cache key
    cacheKeyFn: (key) => (typeof key === "object" ? JSON.stringify(key) : key),
    maxBatchSize: 100,
  });

  if (options?.redis?.prefix) {
    loader.clearRedis = (key) => clearRedis(options?.redis?.prefix, key);
  }

  return {
    loader,
  };
}

/**
 * Will instantiate dataloaders from datasources.
 * This should be done for every incoming GraphQL request
 */
export default function createDataLoaders(uuid, testUser, accessToken) {
  const result = {};
  const track = trackMe().start(uuid);

  result.trackingObject = track;
  const fetchWithConcurrencyLimit = createFetchWithConcurrencyLimit(
    config.fetchConcurrencyLimit
  );
  // Gets a loader by name.
  // A loader will be initialized first time it is called
  function getLoader(name) {
    if (!result[name]) {
      result[name] = setupDataloader(nameToDatasource[name], {
        getLoader,
        track,
        fetch: (url, options) => fetchWithConcurrencyLimit(url, options, name),
        trackingId: uuid,
        testUser,
        accessToken,
      })?.loader;
    }
    return result[name];
  }
  result.getLoader = getLoader;
  return result;
}

/**
 * For testing
 */
export function createMockedDataLoaders() {
  const mockedDatasources = {};
  getFilesRecursive("./src/datasources/mocked")
    .filter(
      (file) =>
        file.path.endsWith("datasource.mocked.js") && require(file.path).load
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.mocked.js", ""),
      load: require(file.path).load,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDatasources[loader.name] = loader;
    });

  mockedDatasources.getLoader = (name) => mockedDatasources[name];

  return mockedDatasources;
}

export function createTestUserDataLoaders() {
  const mockedDataLoaders = createMockedDataLoaders();
  getFilesRecursive("./src/datasources")
    .filter(
      (file) =>
        file.path.endsWith("datasource.js") && require(file.path).testLoad
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.js", ""),
      load: require(file.path).testLoad,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDataLoaders[loader.name] = loader;
    });

  return mockedDataLoaders;
}
