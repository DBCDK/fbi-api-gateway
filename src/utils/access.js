/**
 * @file access util functions
 */

import { filterDuplicateAgencies, resolveManifestation } from "./utils";

/**
 *
 * Check whether any of the users libraries has access to Infomedia.
 * The returned agencyId will be billed for infomedia access.
 *
 * First we check if the users login library has access to info media, if not,
 * we check if any other og the users associated accounts has access to infomedia,
 * if so, we return the first one found.
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
async function infomedia(context) {
  const user = context?.user;

  // get rights from idp
  const idpRights = await context.datasources.getLoader("idp").load("");

  // check if users loggedInAgency has infomedia access
  const loggedInAgencyId = user?.loggedInAgencyId;

  if (loggedInAgencyId && idpRights[loggedInAgencyId]) {
    return loggedInAgencyId;
  }

  /*
   * Alternativly check all users accounts
   */

  // filtered accounts
  const userInfoAccounts = filterDuplicateAgencies(user.agencies);

  const hasAccess = userInfoAccounts?.filter(
    ({ agencyId }) => idpRights[agencyId]
  );

  // check for infomedia access - if any of users agencies subscribes
  if (hasAccess.length > 0) {
    return hasAccess[0]?.agencyId;
  }

  return null;
}

/**
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
export async function getInfomediaAgencyId(context) {
  return await infomedia(context);
}

/**
 *
 * @param {object} context
 *
 * @returns {boolean}
 */
export async function hasInfomediaAccess(context) {
  return !!(await infomedia(context));
}

/**
 * Get type of access for manifestation with given pid.
 *  url access
 *  ill access
 *  infomedia access
 *  digital article service access
 *
 * @param pid
 * @param context
 * @returns {Promise<*>}
 */
export async function resolveAccess(manifestation, context) {
  // We parse the access structure from JED, and convert it
  // to the union type structure.
  // At some point we may choose to follow the structure of JED closely,
  // but it has to be coordinated with stakeholders

  let parent =
    typeof manifestation === "object"
      ? manifestation
      : await resolveManifestation({ pid: manifestation }, context);

  const res = [];

  // use linkchecker to check status of a single accessUrl
  // linkchecker does NOT handle proxy urls, so we skip ebookcentral and ebsco urls
  const linkStatus = async (url) => {
    if (url?.includes("ebookcentral") || url?.includes("ebscohost")) {
      return "OK";
    }

    const linkcheck = await context.datasources
      .getLoader("linkcheck")
      .load({ urls: [url] });
    const link = linkcheck.find((check) => check.url === url);

    return link?.status || "OK";
  };

  parent?.access?.accessUrls?.forEach((entry) => {
    const { proxyUrl, loginRequired } = getProxyUrl(
      entry.url || "",
      context?.user
    );

    res.push({
      __typename: "AccessUrl",
      origin: parseOnlineUrlToOrigin(entry.url),
      url: proxyUrl,
      loginRequired,
      note: entry.note,
      type: entry.type,
      status: linkStatus(proxyUrl) || "OK",
    });
  });

  if (parent?.access?.dbcWebArchive) {
    const archives = await context.datasources
      .getLoader("moreinfoWebarchive")
      .load(parent.pid);

    archives.forEach((archive) => {
      if (archive.url) {
        res.push({
          __typename: "AccessUrl",
          url: archive.url,
          origin: parseOnlineUrlToOrigin(archive.url),
          loginRequired: false,
          type: "RESOURCE",
          status: linkStatus(archive.url) || "OK",
        });
      }
    });
  }

  parent?.access?.ereol?.forEach((entry) => {
    res.push({
      __typename: "Ereol",
      origin: entry.origin,
      url: entry.url,
      canAlwaysBeLoaned: entry.canAlwaysBeLoaned,
      note: entry.note,
    });
  });

  if (parent?.access?.infomediaService?.id) {
    if (
      !["politiken", "jyllands-posten"].some((publication) =>
        parent?.hostPublication?.title?.toLowerCase()?.includes(publication)
      )
    ) {
      res.push({
        __typename: "InfomediaService",
        id: parent.access.infomediaService.id,
      });
    }

    // If this is a omitted online access for politiken or jyllands-posten.
    else {
      const hasPhysical = parent?.accessTypes?.some(
        (t) => (t?.code ?? "").toUpperCase() === "PHYSICAL"
      );

      if (hasPhysical) {
        parent = {
          ...parent,
          access: {
            ...(parent.access ?? {}),
            interLibraryLoanIsPossible: true,
          },
        };
      }
    }
  }

  // Get the issn for this article or periodica
  const issn =
    parent?.access?.digitalArticleService?.issn ||
    parent?.identifiers?.find?.((id) => id.type === "ISSN")?.value;

  if (issn) {
    const journals = await context.datasources
      .getLoader("statsbiblioteketJournals")
      .load("");
    const articleissn = issn.replace(/[^a-z\d]/gi, "");
    const hasJournal = journals && journals[articleissn];
    if (hasJournal) {
      res.push({
        __typename: "DigitalArticleService",
        issn: articleissn,
      });
    }
  }

  if (parent?.access?.interLibraryLoanIsPossible) {
    //if collectionidentifiers includes 870970-accessnew it may be loaned
    const isNewMaterial =
      parent?.collectionIdentifiers?.includes("870970-accessnew");

    const forLoan = checkInterLibraryLoan(parent, context);
    res.push({
      __typename: "InterLibraryLoan",
      loanIsPossible: forLoan,
      accessNew: isNewMaterial,
    });
  }

  // Return array containing all types of access
  return _sortOnlineAccess(res);
}

/**
 * municipality number is the second|third|fourth digit in agencyId
 * @param agencyId
 * @returns {string | undefined}
 */
function parseForMunicipalityNumber(agencyId) {
  return agencyId?.substring(1, 4);
}

/**
 * This one is for ebook.plus - we need to go via a proxy url (if user is logged in)
 * @param url
 * @param user
 * @returns {*}
 */
export function getProxyUrl(url, user) {
  const municipality =
    user?.municipality ||
    parseForMunicipalityNumber(user?.municipalityAgencyId);

  // check if we should proxy this url - for now it is ebookcentral and ebscohost
  const proxyMe =
    url.indexOf("ebookcentral") !== -1 || url.indexOf("ebscohost") !== -1;
  if (proxyMe) {
    // check if user is logged in
    if (user?.userId) {
      const realUrl = `https://bib${municipality}.bibbaser.dk/login?url=${url}`;
      return { proxyUrl: realUrl, loginRequired: proxyMe };
    }
  }

  return { proxyUrl: url, loginRequired: proxyMe };
}

/**
 * Some sources are not for loan (fagbibliografier) - for now a statis list:
 * This method is only called if interLibraryLoan is true - We know that for a fact
 * 159080-fagbib - Besættelsesbibliografien
 * 159081-fagbib - Bibliografi over Dansk Kunst
 * 159082-fagbib - Dansk Historisk Bibliografi
 * 159083-fagbib - Dansk Musiklitterær Bibliografi
 * 159084-fagbib - Kongelige Teater programartikler
 * 159085-fagbib - Dania Polyglotta
 * 159086-fagbib - Sportline
 * @param parent
 * @param context
 */
function checkInterLibraryLoan(parent, context) {
  // This is a blacklist of bases NOT to be loaned
  const notForLoanList = [
    "159080-fagbib",
    "159081-fagbib",
    "159082-fagbib",
    "159083-fagbib",
    "159084-fagbib",
    "159085-fagbib",
    "159086-fagbib",
  ];
  // we check on objectId - first part is the source
  const source = parent?.objectId?.split(":")[0];
  return !notForLoanList.includes(source);
}

/**
 * Get domain from url
 * @param url
 * @return String
 *  A parsed string eg. "DBC Webarkiv" or name of host eg. "infolink2003.elbo.dk"
 *
 * */
export function parseOnlineUrlToOrigin(url) {
  try {
    const parsedUrl = new URL(url);
    if (parsedUrl["host"] === "moreinfo.addi.dk") {
      return "DBC Webarkiv";
    } else {
      return (parsedUrl["host"] && parsedUrl["host"]) || "";
    }
  } catch (e) {
    return "";
  }
}

/**
 * Handle special cases - for now we want filmstriben/fjernleje BEFORE filmstriben/biblioteket AND
 * we want dansk film institut (dfi) to come last - it is not a 'real' url but a link to a description
 * @param onlineAccess
 * @return {*}
 * @private
 */
function _sortOnlineAccess(onlineAccess) {
  const specialSort = (a, b) => {
    // fjernleje should be on top
    if (b.url && b.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return 1;
    } else if (a.url && a.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return -1;
      // dfi is not a 'real' online url - sort low
    } else if (b.url && b.url.indexOf("dfi.dk") !== -1) {
      return -1;
    } else if (a.url && a.url.indexOf("dfi.dk") !== -1) {
      return 1;
    }
    return 0;
  };

  return onlineAccess.sort(specialSort);
}
