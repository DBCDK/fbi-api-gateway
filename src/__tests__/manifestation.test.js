import { createMockedDataLoaders } from "../datasourceLoader";
import { performTestQuery } from "../utils/utils";

describe("Manifestation", () => {
  test("fetch by pid", async () => {
    const result = await performTestQuery({
      query: `query ($pid: String!) {
        manifestation(pid: $pid) {
          pid 
        }
      }`,
      variables: { pid: "870970-basis:26521556" },
      context: { datasources: createMockedDataLoaders() },
    });
    expect(result).toMatchSnapshot();
  });
  test("fetch by faust", async () => {
    const result = await performTestQuery({
      query: `query ($faust: String!) {
        manifestation(faust: $faust) {
          pid 
        }
      }`,
      variables: { faust: "26521556" },
      context: { datasources: createMockedDataLoaders() },
    });
    expect(result).toMatchSnapshot();
  });
  test("fetch multi by pid", async () => {
    const result = await performTestQuery({
      query: `query ($pid: [String!]!) {
        manifestations(pid: $pid) {
          pid 
        }
      }`,
      variables: { pid: ["870970-basis:26521556"] },
      context: { datasources: createMockedDataLoaders() },
    });
    expect(result).toMatchSnapshot();
  });
  test("fetch multi by faust", async () => {
    const result = await performTestQuery({
      query: `query ($faust: [String!]!) {
        manifestations(faust: $faust) {
          pid 
        }
      }`,
      variables: { faust: ["26521556"] },
      context: { datasources: createMockedDataLoaders() },
    });
    expect(result).toMatchSnapshot();
  });
  test("Query all fields", async () => {
    const result = await performTestQuery({
      query: `query ($faust: String!) {
        manifestation(faust: $faust) {
          pid
          cover {
            thumbnail
            detail_117
            detail_207
            detail_42
            detail_500
            detail
          }
          titles {
            main
            full
            sort
            alternative
            identifyingAddition
            original
            parallel
            standard
            translated
          }
          abstract
          accessTypes {
            display
            code
          }
          access {
            __typename
            ... on AccessUrl {
              origin
              url
            }
            ... on Ereol {
              origin
              url
              canAlwaysBeLoaned
            }
            ... on InterLibraryLoan {
              loanIsPossible
            }
            ... on InfomediaService {
              id
            }
            ... on DigitalArticleService {
              issn
            }
          }
          audience {
            generalAudience
            ages {
              begin
              end
              display
            }
            libraryRecommendation
            childrenOrAdults {
              display
              code
            }
            schoolUse {
              display
              code
            }
            primaryTarget
            let
            lix
          }
          contributors {
            display
            nameSort
          }
          contributorsFromDescription
          creators {
            display
            nameSort
          }
          creatorsFromDescription
          classifications {
            code
            dk5Heading
            display
            entryType
            system
          }
          dateFirstEdition {
            display
            year
            endYear
            frequency
          }
          edition {
            summary
            edition
            contributors
            publicationYear {
              display
              year
              endYear
              frequency
            }
          }
          latestPrinting {
            summary
            printing
            publicationYear {
              display
            }
          }
          genreAndForm
          hostPublication {
            title
            creator
            issn
            isbn
            issue
            notes
            pages
            publisher
            
            year {
              display
            }
            summary
          }
          identifiers {
            type
            value
          }
          languages {
            notes
            main {
              display
              isoCode
            }
            original {
              display
              isoCode
            }
            parallel {
              display
              isoCode
            }
            spoken {
              display
              isoCode
            }
            subtitles {
              display
              isoCode
            }
            abstract {
              display
              isoCode
            }
          }
          manifestationParts {
            heading
            parts {
              title
              creators {
                display
              }
              classifications {
                code
              }
              subjects {
                display
              }
              creatorsFromDescription
              playingTime
            }
            type
          }
          materialTypes {
            general
            specific
          }
          notes {
            type
            heading
            display
          }
          ownerWork {
            workId
          }
          relatedPublications {
            heading
            title
            issn
            isbn
            urlText
            url
          }
          physicalDescriptions {
            summary
            accompanyingMaterial
            additionalDescription
            extent
            numberOfPages
            numberOfUnits
            playingTime
            requirements
            size
            technicalInformation
          }
          publisher
          
          shelfmark {
            postfix
            shelfmark
          }
          source
          subjects {
            all {
              display
            }
            dbcVerified {
              display
              ... on SubjectText {
                language {
                  display
                  isoCode
                }
              }
            }
          }
          volume
          tableOfContents {
            heading
            content
            listOfContent {
              content
            }
          }
        }
      }`,
      variables: { faust: "26521556" },
      context: { datasources: createMockedDataLoaders() },
    });
    expect(result).toMatchSnapshot();
  });
});
