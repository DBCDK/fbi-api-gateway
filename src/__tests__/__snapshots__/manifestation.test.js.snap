// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manifestation Query all fields 1`] = `
Object {
  "data": Object {
    "manifestation": Object {
      "abstract": Array [
        "Dummy Some abstract ...",
      ],
      "access": Array [
        Object {
          "__typename": "AccessUrl",
          "origin": "infolink2003.elbo.dk",
          "url": "http://infolink2003.elbo.dk/DVT/dokumenter/doc/15271.pdf",
        },
        Object {
          "__typename": "DigitalArticleService",
          "issn": "09021620",
        },
      ],
      "accessTypes": Array [
        Object {
          "code": "PHYSICAL",
          "display": "Dummy fysisk",
        },
        Object {
          "code": "ONLINE",
          "display": "Dummy online",
        },
      ],
      "audience": Object {
        "ages": Array [
          Object {
            "begin": 10,
            "display": "Dummy 10-14",
            "end": 14,
          },
        ],
        "childrenOrAdults": Array [
          Object {
            "code": "FOR_CHILDREN",
            "display": "Dummy til børn",
          },
        ],
        "generalAudience": Array [
          "Dummy general audience",
        ],
        "let": "Dummy some let",
        "libraryRecommendation": "Dummy some library recommendation",
        "lix": "Dummy some lix",
        "primaryTarget": Array [
          "Dummy Some primary target",
        ],
        "schoolUse": Array [
          Object {
            "code": "FOR_SCHOOL_USE",
            "display": "Dummy til skolebrug",
          },
        ],
      },
      "catalogueCodes": Object {
        "nationalBibliography": Array [
          "national_bibliography",
        ],
        "otherCatalogues": Array [
          "other_catalogueCodes",
          "even_other_catalogueCodes",
        ],
      },
      "classifications": Array [
        Object {
          "code": "Dummy 86-096",
          "display": "Dummy 86-096, Skønlitteratur",
          "dk5Heading": "Dummy Skønlitteratur",
          "entryType": "NATIONAL_BIBLIOGRAPHY_ENTRY",
          "system": "Dummy DK5",
        },
        Object {
          "code": "Dummy 86-096 additional",
          "display": "Dummy 86-096, Skønlitteratur additional",
          "dk5Heading": "Dummy Skønlitteratur additional",
          "entryType": "NATIONAL_BIBLIOGRAPHY_ADDITIONAL_ENTRY",
          "system": "Dummy DK5 additional",
        },
      ],
      "contributors": Array [
        Object {
          "display": "Dummy Jens Jensen",
          "nameSort": "Dummy Jensen Jens",
        },
      ],
      "contributorsFromDescription": Array [
        "Dummy på dansk ved Vivi Berendt",
      ],
      "cover": Object {
        "detail": "link-to-detail",
        "detail_117": "link-to-detail_117",
        "detail_207": "link-to-detail_207",
        "detail_42": "link-to-detail_42",
        "detail_500": "link-to-detail_500",
        "thumbnail": "link-to-thumbnail",
      },
      "creators": Array [
        Object {
          "display": "Dummy Jens Jensen",
          "nameSort": "Dummy Jensen Jens",
        },
        Object {
          "display": "Dummy Some Corporation",
          "nameSort": "Dummy Some Corporation for sorting",
        },
      ],
      "creatorsFromDescription": Array [
        "Dummy tekst af William Warren",
      ],
      "dateFirstEdition": Object {
        "display": "Dummy first edition",
        "endYear": 1950,
        "frequency": "1",
        "year": 1950,
      },
      "edition": Object {
        "contributors": Array [],
        "edition": "Dummy 3. i.e. 2 udgave",
        "publicationYear": Object {
          "display": "Dummy 2005",
          "endYear": null,
          "frequency": null,
          "year": 2005,
        },
        "summary": "Dummy 3. i.e. 2 udgave, 2005",
      },
      "genreAndForm": Array [
        "Dummy some genre",
      ],
      "hostPublication": Object {
        "creator": "Dummy Some Creator",
        "isbn": "Dummy some isbn",
        "issn": "Dummy 1395-7961",
        "issue": "Dummy some issue",
        "notes": Array [
          "Dummy a note",
        ],
        "pages": "Dummy 140-145",
        "publisher": "Dummy Some Publisher",
        "summary": "Dummy Årsskrift / Carlsbergfondet, 2006",
        "title": "Dummy Årsskrift / Carlsbergfondet",
        "year": Object {
          "display": "Dummy 2006",
        },
      },
      "identifiers": Array [
        Object {
          "type": "ISBN",
          "value": "Dummy 1234567891234",
        },
      ],
      "languages": Object {
        "abstract": Array [
          Object {
            "display": "Dummy dansk",
            "isoCode": "Dummy dan",
          },
        ],
        "main": Array [
          Object {
            "display": "Dummy dansk",
            "isoCode": "Dummy dan",
          },
        ],
        "notes": Array [
          "Dummy dansk",
          "dummy english",
          "dummy german",
        ],
        "original": Array [
          Object {
            "display": "Dummy dansk",
            "isoCode": "Dummy dan",
          },
        ],
        "parallel": Array [
          Object {
            "display": "Dummy dansk",
            "isoCode": "Dummy dan",
          },
        ],
        "spoken": Array [
          Object {
            "display": "Dummy dansk",
            "isoCode": "Dummy dan",
          },
        ],
        "subtitles": Array [
          Object {
            "display": "Dummy dansk",
            "isoCode": "Dummy dan",
          },
        ],
      },
      "latestPrinting": Object {
        "printing": "Dummy 11. oplag",
        "publicationYear": Object {
          "display": "Dummy 2020",
        },
        "summary": "Dummy 11. oplag, 2020",
      },
      "manifestationParts": Object {
        "heading": "Dummy Indhold:",
        "parts": Array [
          Object {
            "classifications": Array [
              Object {
                "code": "Dummy 78.424",
              },
            ],
            "creators": Array [
              Object {
                "display": "Dummy H. C. Lumbye",
              },
            ],
            "creatorsFromDescription": Array [
              "Dummy arr.: Peter Ettrup Larsen",
            ],
            "playingTime": "Dummy playing time",
            "subjects": null,
            "title": "Dummy Bouquet royal",
          },
          Object {
            "classifications": Array [
              Object {
                "code": "Dummy 88.424",
              },
            ],
            "creators": Array [
              Object {
                "display": "Dummy A. C. Handerson",
              },
            ],
            "creatorsFromDescription": Array [
              "Dummy arr.: Don Alderson Drumf",
            ],
            "playingTime": "Dummy long playing time",
            "subjects": null,
            "title": "Dummy Banker Loyal",
          },
        ],
        "type": "MUSIC_TRACKS",
      },
      "materialTypes": Array [
        Object {
          "general": "Dummy general type",
          "specific": "Dummy bog",
        },
      ],
      "notes": Array [
        Object {
          "display": Array [
            "Dummy Indspillet i Ꜳrhus Musikhus 12.-14. juni 2020",
          ],
          "heading": null,
          "type": "NOT_SPECIFIED",
        },
      ],
      "ownerWork": Object {
        "workId": "work-of:870970-basis:26521556",
      },
      "physicalDescriptions": Array [
        Object {
          "accompanyingMaterial": null,
          "additionalDescription": null,
          "extent": "Dummy 1 dvd-rom",
          "numberOfPages": null,
          "numberOfUnits": null,
          "playingTime": null,
          "requirements": "Dummy Nødvendigt udstyr Xbox One. - Med multiplayerfunktion",
          "size": null,
          "summary": "Dummy 1 dvd-rom Xbox One Nødvendigt udstyr Xbox One. - Med multiplayerfunktion Spiludvikler fremgår ikke af materialet",
          "technicalInformation": "Dummy Xbox One",
        },
      ],
      "pid": "870970-basis:26521556",
      "publisher": Array [
        "Dummy Lægeforeningen",
      ],
      "relatedPublications": Array [
        Object {
          "heading": "Dummy Tidligere titel:",
          "isbn": null,
          "issn": "Dummy 0105-0508",
          "title": Array [
            "Dummy Yngre læger",
          ],
          "url": null,
          "urlText": null,
        },
        Object {
          "heading": "Dummy Udgave i andet medium: Også på cd-rom",
          "isbn": null,
          "issn": "Dummy 1399-4174",
          "title": Array [
            "Dummy Ugeskrift for læger",
          ],
          "url": null,
          "urlText": null,
        },
      ],
      "series": Array [
        Object {
          "numberInSeries": Object {
            "number": Array [
              1,
            ],
          },
          "title": "De syv søstre-serien",
        },
      ],
      "shelfmark": Object {
        "postfix": "Dummy some postfix",
        "shelfmark": "Dummy some shelfmark",
      },
      "source": Array [
        "Dummy some source",
      ],
      "subjects": Object {
        "all": Array [
          Object {
            "display": "Dummy Some fictional subject",
          },
          Object {
            "display": "Dummy 1950-1980",
          },
          Object {
            "display": "Dummy Jens Jensen",
          },
          Object {
            "display": "Dummy Some Corporation",
          },
        ],
        "dbcVerified": Array [
          Object {
            "display": "Dummy Some fictional subject",
            "language": Object {
              "display": "Dummy dansk",
              "isoCode": "dum dan",
            },
          },
          Object {
            "display": "Dummy 1950-1980",
          },
          Object {
            "display": "Dummy Jens Jensen",
          },
          Object {
            "display": "Dummy Some Corporation",
          },
        ],
      },
      "tableOfContents": Object {
        "content": null,
        "heading": "Dummy Indhold",
        "listOfContent": Array [
          Object {
            "content": null,
          },
          Object {
            "content": "Dummy Tykke-Olsen m.fl.",
          },
          Object {
            "content": "Dummy Over skulderen",
          },
        ],
      },
      "titles": Object {
        "alternative": Array [
          "Dummy Some Title: Alternative",
        ],
        "full": Array [
          "Dummy Some Title: Full",
        ],
        "identifyingAddition": "Dummy Indlæst af Jens Jensen",
        "main": Array [
          "Dummy Some Title",
        ],
        "original": Array [
          "Dummy Some Title: Original",
        ],
        "parallel": Array [
          "Dummy Parallel Title 1",
          "Parallel Title 2",
        ],
        "sort": "Dummy Some Title Sort",
        "standard": "Dummy Some Title Standard",
        "translated": Array [
          "Dummy En Oversat Titel",
        ],
      },
      "volume": "Dummy Bind 2",
    },
  },
}
`;

exports[`Manifestation fetch by faust 1`] = `
Object {
  "data": Object {
    "manifestation": Object {
      "pid": "870970-basis:26521556",
    },
  },
}
`;

exports[`Manifestation fetch by pid 1`] = `
Object {
  "data": Object {
    "manifestation": Object {
      "pid": "870970-basis:26521556",
    },
  },
}
`;

exports[`Manifestation fetch multi by faust 1`] = `
Object {
  "data": Object {
    "manifestations": Array [
      Object {
        "pid": "870970-basis:26521556",
      },
    ],
  },
}
`;

exports[`Manifestation fetch multi by pid 1`] = `
Object {
  "data": Object {
    "manifestations": Array [
      Object {
        "pid": "870970-basis:26521556",
      },
    ],
  },
}
`;
